### PR TITLE
Changes needed for SUSY parameter scans

### PR DESCRIPTION
**Pull request by bendavid**
Adds the possibility to provide a list of configurations to Pythia8 and randomly select from the list of possible configurations at beginLuminosityBlock time.

This allows to produce a parameter scan in a single monte carlo request/dataset for whatever can be varied at the pythia level.  This is targeted initially for SUSY fastsim signal production in 80x where the parameter scan is over the SUSY decay chain defined by the SLHA header, which can be inlined as part of the pythia8 configuration.

This is the (near) final implementation of the proposal discussed in https://indico.cern.ch/event/487292/contribution/15/attachments/1218978/1781280/susyscan-Jan28-2016.pdf

n.b.  I consider that this is not fully tested until the changes in https://github.com/cms-sw/cmsdist/pull/2131 are in the IB and these can be properly tested together.

### CODE REVIEW

File: GeneratorInterface/Core/src/BaseHadronizer.cc [hunk 1]
```diff
@@ -17,4 +31,28 @@ BaseHadronizer::BaseHadronizer( edm::ParameterSet const& ps )
 
 }
 
+void BaseHadronizer::randomizeIndex(edm::LuminosityBlock const& lumi, CLHEP::HepRandomEngine* rengine) {
+  if (randomInitWeights_.size()>0) {
+    //randomly select from a list of provided configuration sets (for parameter scans)
+    
+    //seeds std 32-bit mersene twister with 50 random integers plus run and lumi section numbers
```
Comment (bendavid) [inline GeneratorInterface/Core/src/BaseHadronizer.cc]: @Dr15Jones this is a further updated implementation on how to generate random numbers deterministically in beginLuminosityBlock.  In this case an independent mersene twister generator is initialized using random numbers from the CMSSW random number engine combined with the run and lumi section numbers.

**Comment by bendavid** [comment]
The DataFormats changes are needed to store in the event information on which config has been chosen.  (This information is duplicated in both the lumi and event for user convenience, and because the GenLumiInfoProduct is not available until the endLuminosityBlock in case things are being done on the fly)

**Comment by bendavid** [comment]
Doing this safely requires a more complete re-initizalization of pythia at lumi section boundaries (pointer to main program and those related to userhooks are completely deleted and recreated rather than just calling init() again)

I've tested that this does not leak memory at any observable rate over several thousand lumi sections.

The inheritance of the Pythia8 interfaces was also slightly reorganized so that everything really inherits from BaseHadronizer now.  (previously some of the particle guns did not, which made it more painful to extend the hadronizer api)

**Comment by bendavid** [comment]
Note that for the purposes of verifying that these changes don't break anything, it is rather unfortunate that the issue of lumi transitions in relvals was not addressed.  (As far as I know, relvals are still running strictly one lumi-section per GEN or GEN-SIM job, such that anything which breaks at lumi-section boundaries goes completely untested)

**Comment by govoni** [comment]
+1

**Comment by covarell** [comment]
Is there a specific reason why the implementation is done by touching the very basic Hadronizer/Generator classes and not defining new ones derived from those where the randomization is implemented?

**Comment by bendavid** [comment]
Yes, the idea was to leave open the possibility to use this also for other generators in the future.

The new code is only called if the new configuration options ("RandomizedParameters") are used.

**Comment by covarell** [comment]
OK, this makes sense

**Comment by bendavid** [comment]
(this has conflicts and needs to be rebased)

**Comment by bendavid** [comment]
due to scope creep and associated delay this is replaced for now with https://github.com/cms-sw/cmssw/pull/13530 for 81x.  A new pull request for backporting to 80x will be created later.

**Pull request closed without merging**